README.MD
# Thiago do Carmo Nunes - Economics Master Degree

## Author : 
THIAGO DO CARMO NUNES

## Date : 
17JUL2022

# Program:
- Master's Defense Exam - Postgraduate Program in Economics  | Exame de Defesa de Mestrado - Programa de Pós-Graduação em Economia
- FEDERAL UNIVERSITY OF ABC \ UNIVERSIDADE FEDERAL DO ABC (Alameda da Universidade, s/nº - Bairro Anchieta - São Bernardo do Campo - SP - Brasil - CEP: 09606-045 , Brasil )

# Title: 
COMPARISON OF THE PREDICTIVE PERFORMANCE OF ARIMA, VAR AND VEC MODELS IN COINTEGRATED TIME SERIES

# Abstract:

This work aims to compare the predictive power of the univariate ARIMA models with the multivariate VAR and VEC models under the cointegration hypothesis. The analysis was condutec on Monte Carlo simulations using statistical methods to evaluate the predictive power. A model was estimated using exchange rate (USD/BRL) and 12-month cumulative inflation (IPCA) to check the Monte Carlo simulation insights.

The model's predictive power on out-of-sample data is similar to Monte Carlo simulations' 95% confidence intervals. The conclusion of the exchange rate and inflation data experiment was in line with simulations. The predictive power of the models was similar within the confidence intervals of the mean. However, the size of the confidence intervals demands more complex validations before choosing the modeling specification for prediction.

The results are also in line with market best practices, where even knowing the appropriate theoretical model specification, we should pay attention to out-of-sample performance experiments for greater assertiveness in predictions.

The work concluded that even in a set of cointegrated series, where there is a guide for modeling VEC in the econometric literature, to get more assertiveness of predictions, it is necessary to do out-of-sample experiments outside the sample to choose the model with greater predictive power.

# Titulo: 
COMPARAÇÃO DA PERFORMANCE PREDITIVA DE MODELOS ARIMA, VAR E VEC EM SÉRIES TEMPORAIS COINTEGRADAS

# Resumo: 

O objetivo desta dissertação é comparar a capacidade preditiva dos modelos univariados ARIMA com os modelos multivariados VAR e VEC sob a hipótese de cointegração. A análise foi baseada em simulações de Monte Carlo utilizando métricas estatísticas para avaliar a performance preditiva dos modelos. Para validar as conclusões obtidas via simulações foi realizado um experimento de modelagem com séries temporais reais de cambio (USD/BRL) e inflação acumulada em 12 meses (IPCA).

As simulações mostraram que a capacidade preditiva dos modelos em dados fora da amostra de treinamento é em média equivalente dentro dos intervalos de confiança de 95%. No experimento com dados reais a capacidade preditiva dos modelos foi similar dentro dos intervalos de confiança da média, alinhado a conclusão obtida via simulações.

Os resultados também se mostraram alinhados com as práticas de mercado onde mesmo sabendo a especificação teórica mais adequada ao conjunto de dados no treinamento, deve-se atentar para experimentos de performance fora da amostra para maior assertividade nas predições. Contudo, a amplitude dos intervalos de confiança aponta para a necessidade de realizar validações mais complexas previamente à escolha da estrutura de modelagem para predição.

A conclusão do trabalho foi que mesmo em um conjunto de séries cointegradas, onde pela literatura econométrica existe uma orientação para modelagem via VEC, para maior assertividade das predições é necessário fazer experimentos de validação fora da amostra para escolher o modelo com maior capacidade preditiva.


CHANGELOG.md
# Changelog

Esse arquivo contém as principais alterações feitas na dissertação de acordo com os feedbacks recebidos pela banca.


# Ações

## Reduzir o esforço cognitivo do Leitor
- Alterada a introdução para abordar: Porque, Como vamos Fazer e as premissas.
- Resumo e Objetivo bem direto ao ponto somente com as principais conclusões
- Linha do tempo dos modelos que estava na introdução foi migrada para a primeira parte da metodologia.
- Sessões com subtópicos para ficar claro ao leitor o que ele vai ler
- Quebra da sessão de metodologia em (i) mostrar uma linha do tempo mostrando a evolução dos modelos escolhidos como alvo desse trabalho, (ii) o design e especificações mais técnicas dos modelos e (iii) a metodologia de simulação e (iv) as estratégias de validação dos modelos.
- Removi as tabelas que estavam mais complexas de entender e deixei somente as que tem um resultado mais interessante pra ser discutido no texto.

## Otimização de Hiperparametros
- Incluído critério BIC no texto.
- 

## Revisão Bibliográfica
- Incluído o que a literatura diz sobre esses modelos? (Vec performa melhor em que situações…)


## Fazer uma aplicação com dados reais
- Aplicação Cambio e Inflação

## Gráficos:
- Ajuste dos gráficos removendo barras.



DM_test.R
path <- "C:/Users/thiag/Documents/1 -Mestrado/SprintFinal/Results/DM/"

filenames <- list.files(path, pattern="*rds")

DM_list <- list()

for (file in filenames) {
  
  readed_DM <- readRDS(paste0(path,file), refhook = NULL)
  
  base_split <- strsplit(file, split = "&")[[1]]
  par__nobs <- base:::strsplit(base_split[2], split = "__")[[1]][2]
  par__n_ma <- base:::strsplit(base_split[3], split = "__")[[1]][2]
  par__n_fold_ts <- base:::strsplit(base_split[4], split = "__")[[1]][2]
  par__n_ts <- substr(base:::strsplit(base_split[5], split = "__")[[1]][2],1,1)

  
  DM_list[[file]] <- list("file"=file,"n_obs"=par__nobs,"n_ma"=par__n_ma,"n_fold_ts"=par__n_fold_ts,"n_ts"=par__n_ts,"SAME_1_pct"=mean(readed_DM>0.01),"SAME_5_pct"=mean(readed_DM>0.05),"SAME_10_pct"=mean(readed_DM>0.1))
  
}

DM_consolidado <- do.call(rbind.data.frame, DM_list)

writexl::write_xlsx(DM_consolidado,"C:/Users/thiag/Documents/1 -Mestrado/SprintFinal/Results/consolidado/DM.xlsx")

ecdf_chart.R
library(lattice)
library(latticeExtra)
library(ggplot2)
library(reshape2)
library(boot)

err <- "MAE"

setwd("C:/Users/thiag/Documents/1 -Mestrado/SprintFinal/Defesa/Results/")

path <- paste0(err,"/")

filenames <- list.files(path, pattern="*rds")

for (file in filenames) {
  
  readed_file <- readRDS(paste0(path,file), refhook = NULL)
  readed_file$id <- rownames(readed_file)
  
  readed_file_melted <- melt(readed_file[,c("id","ARIMA","VAR","VEC")], id="id",value.name= err)
  
  # ECDF PLOT
  title <- sub('\\.rds$', '', file)
  print(title)

  png(file=paste0("consolidado/ECDF_charts/",title,".png"),width=1920, height=1080)
  
  
  p <- ggplot(readed_file_melted, aes(MAE, colour = variable)) + stat_ecdf()
  p <- p + scale_color_manual(values=c("blue","yellow","black"))
  p <- p + labs(y="ECDF",title=title) 
  

  plot(p)
  
  dev.off()

  
  # ks.test(resultados_df$ARIMA,resultados_df$VAR,alternative = c("less"))
  # ks.test(resultados_df$ARIMA,resultados_df$VEC,alternative = c("less"))
  # ks.test(resultados_df$VAR,resultados_df$VEC,alternative = c("less"))
  
  
}


CointegratedCode.R
# Libraries
library(stats)
require(tsDyn)
library(vars)
library(Metrics)
library(data.table)
library(dplyr)
library(progress)
library(urca)
library(tseries)
library(forecast)
library(DescTools)
require(utils)

# Simulation Params Grid
grid_params <- expand.grid(
  arima_n_obs = c(500,250, 100),
  n_ma  = c(0, 12),
  n_fold_ts = c(0.85,0.9, 0.95),
  n_ts = c(3,2)
)

seed <- 2022
ite_total <- 1000

setwd("C:/Users/thiag/Documents/1 -Mestrado/SprintFinal/Defesa/Results/")
dir.create("MAE")
dir.create("MSE")
dir.create("DM")
dir.create("consolidado")
dir.create("consolidado/ECDF_charts")
dir.create("consolidado/MAE")
dir.create("consolidado/MSE")
dir.create("consolidado/MAE/boot_charts")
dir.create("consolidado/MSE/boot_charts")


# Loop
for (grid_i in c(1:nrow(grid_params))) {

  selecionados <- grid_params[grid_i, ]

  arima_n_obs <- selecionados[["arima_n_obs"]]
  plot <- F
  n_ma <- selecionados[["n_ma"]]
  n_fold_ts <- selecionados[["n_fold_ts"]]
  n_ts <- selecionados[["n_ts"]]

  label_to_save <- paste0(
    "n_obs__", arima_n_obs, " & ",
    "n_ma__", n_ma, " & ",
    "n_fold_ts__", n_fold_ts * 100, " & ",
    "n_ts__", n_ts)

  progress_general <- round(grid_i/nrow(grid_params),2)
  pb <- progress_bar$new(
    format = paste0("  ", progress_general, "  simulating [:bar] :percent eta: :eta"),
    total = ite_total, clear = FALSE, width= 60)

  results_list <- list()
  results_list_mse <- list()

  diebold_mariano_list_ARIMA <- list()
  diebold_mariano_list_VAR <- list()
  diebold_mariano_list_VAR_ARIMA <- list()
  df_list <- list()
  
  for (i in c(1:ite_total)) {
    
    pb$tick()
    
    seed <- seed + 1
    set.seed(seed)

    # Generate DataFrame
    sd <- 0.8
    cointegration_vector <- c(
      1,
      rnorm(1, 2, 1),
      rnorm(1, 2, 1))
  
    y3 <- stats::arima.sim(
      list(
        order = c(3, 1, 2),
        ar = c(0.6, -0.4, -0.3),
        ma = c(-0.3, 0.2),
        sd = sd),
      n = arima_n_obs)
  
    y2 <- stats::arima.sim(
      list(
        order = c(3, 1, 2),
        ar = c(0.5, -0.3, -0.4),
        ma = c(-0.2, -0.25),
        sd = sd),
      n = arima_n_obs)
  
    u <- stats::arima.sim(
      list(
        order = c(1, 0, 0),
        sd = sd, 
        ar = 0.2),
        n = (arima_n_obs + 1))
  
    y1 <- u + cointegration_vector[2] * y2 + cointegration_vector[3] * y3
  
    df <- as.data.frame(scale(cbind(y1, y2, y3))) %>% as.data.table()
  
    df___diff <- df %>% mutate_each(funs(. - lag(.))) %>% as.data.table()
  
    jotest <- ca.jo(df, type = "eigen")
    resultado_coint <- sum(jotest@cval[, 1] < jotest@teststat)
  
    if (resultado_coint == 0) {
      print("pulou coint")
      next
    }
    if (adf.test(df$y1)$p.value <= 0.05 | adf.test(df$y2)$p.value <= 0.05 | adf.test(df$y3)$p.value <= 0.05) {
      print("pulou adf")
      next
    }
    if (adf.test(diff(df$y1))$p.value > 0.05 | adf.test(diff(df$y2))$p.value > 0.05 | adf.test(diff(df$y3))$p.value > 0.05) {
      print("pulou o adf diff")
      # next
    }
  
    if (n_ma != 0) {
      df_ma <- as.data.table(na.omit(forecast::ma(df,n_ma)))
      colnames(df_ma) <- c("y1","y2","y3")
      df <- df_ma
    }
  
    if (n_ts == 2) {
      df <- df[,1:2]
    }
  
    n <- nrow(df)
    fold_time <- floor(n * n_fold_ts)
  
    train__ <- df[1:fold_time]
    train__diff <- df___diff[2:fold_time]
    test__ <- df[(fold_time + 1):n]
    n_ahead <- nrow(test__)
    last_obs_value <- as.numeric(train__[fold_time, "y1"])
  
    # Models
    modelo___arima <- forecast::auto.arima(
      x = train__$y1, allowdrift = F, method = "ML",d = 1)
  
    # Best Model by BIC
    BIC_var_list <- list()
    BIC_vec_list <- list()
  
    for (order_var in c(1:10)) {
      var_i <- vars::VAR(train__diff, p = order_var)
      vec_i <- tsDyn::VECM(train__,
        lag = order_var, estim = "ML",
        include = c("const"), r = 1)
  
      BIC_var_list[[order_var]] <- BIC(var_i)
      BIC_vec_list[[order_var]] <- BIC(vec_i)
    }
  
    ## Selected order for VAR and VEC
    var_order <- which.min(unlist(BIC_var_list))
    vec_order <- which.min(unlist(BIC_vec_list))
  
    #VAR
    modelo___VAR <- vars::VAR(train__diff, p = var_order)
    #VECM
    modelo___VEC <- tsDyn::VECM(train__, lag = vec_order, estim = "ML",
      include = c("const"), r = 1)
  
    # Prediction
    ## ARIMA
    modelo___arima_predict <- as.numeric(
      predict(modelo___arima, n.ahead = n_ahead)$pred)
    ## VAR
    modelo___VAR_predict_diff <- as.numeric(
      predict(modelo___VAR, n.ahead = n_ahead)$fcst$y1[, "fcst"])
    modelo___VAR_predict <- last_obs_value + cumsum(modelo___VAR_predict_diff)
    ## VEC
    modelo___VEC_predict <- as.numeric(
      predict(modelo___VEC, n.ahead=n_ahead)[, "y1"])
  
    ## Consolidated
    prediction__df <- as.data.table(
      cbind(test__$y1, modelo___arima_predict,
        modelo___VAR_predict, modelo___VEC_predict))
  
    # Obs_Test
    y_obs_test <- test__$y1
  
    residuals_ARIMA <- y_obs_test - modelo___arima_predict
    residuals_VAR <- y_obs_test - modelo___VAR_predict
    residuals_VEC <- y_obs_test - modelo___VEC_predict
  
    diebold_mariano_test_VEC_ARIMA <-  dm.test(residuals_ARIMA, residuals_VEC)
    diebold_mariano_test_VEC_VAR <-  dm.test(residuals_VAR, residuals_VEC)
    diebold_mariano_test_VAR_ARIMA <- dm.test(residuals_VAR, residuals_ARIMA)
  
    diebold_mariano_list_ARIMA <- append(diebold_mariano_list_ARIMA,
      diebold_mariano_test_VEC_ARIMA$p.value)
    diebold_mariano_list_VAR <- append(diebold_mariano_list_VAR,
      diebold_mariano_test_VEC_VAR$p.value)
    diebold_mariano_list_VAR_ARIMA <- append(diebold_mariano_list_VAR_ARIMA,
      diebold_mariano_test_VAR_ARIMA$p.value)
  
    #Metrics
    ## MAE
    modelo___arima_predict___MAE <- mae(y_obs_test, modelo___arima_predict)
    modelo___VAR_predict___MAE <- mae(y_obs_test, modelo___VAR_predict)
    modelo___VEC_predict___MAE <- mae(y_obs_test, modelo___VEC_predict)
  
    resultados <- as.data.frame(
      cbind("ARIMA"=modelo___arima_predict___MAE,"VAR"=modelo___VAR_predict___MAE,
        "VEC"=modelo___VEC_predict___MAE))
  
    resultados$BEST_ALL <- as.matrix(apply(
      resultados[,c("ARIMA","VAR","VEC")], 1, which.min))
  
    results_list[[i]] <- resultados
    ## MSE
    modelo___arima_predict___MSE <- mse(y_obs_test, modelo___arima_predict)
    modelo___VAR_predict___MSE <- mse(y_obs_test, modelo___VAR_predict)
    modelo___VEC_predict___MSE <- mse(y_obs_test, modelo___VEC_predict)
  
    resultados_MSE <- as.data.frame(cbind("ARIMA" = modelo___arima_predict___MSE,
      "VAR" = modelo___VAR_predict___MSE, "VEC" = modelo___VEC_predict___MSE))
    resultados_MSE$BEST_ALL <- as.matrix(apply(
      resultados_MSE[,c("ARIMA", "VAR", "VEC")], 1, which.min))
  
    results_list_mse[[i]] <- resultados_MSE
  
    #Plots
    if (plot) {
      plot.ts(df$y1, pch = 19,col = "black")
      len_train <- train__$y1
      lines(c(rep(NA,length(len_train)),modelo___arima_predict),type = "b", pch = 16,col = "blue")
      lines(c(rep(NA,length(len_train)),modelo___VAR_predict),type = "b", pch = 17,col = "red")
      lines(c(rep(NA,length(len_train)),modelo___VEC_predict),type = "b", pch = 18,col = "green")
      legend("topleft",
             legend=c("actuals","ARIMA","VAR","VECM"),
             col=c("black","blue","red","green"),
             lty = 1:2, cex=0.8)
      round_metrics = 3
      Metrics_Comp <- paste0("ARIMA(",round(modelo___arima_predict___MAE,round_metrics),") | ","VAR(",round(modelo___VAR_predict___MAE,round_metrics),") | ","VEC(",round(modelo___VEC_predict___MAE,3),")")
      
      title(paste("i :",i," |","MAE: ",Metrics_Comp))
    }

  }
  
  ## MAE
  resultados_df <- do.call(rbind.data.frame, results_list)
  table(resultados_df$BEST_ALL)
  ## MSE
  resultados_df_MSE <- do.call(rbind.data.frame, results_list_mse)
  table(resultados_df_MSE$BEST_ALL)
  ## Diebold Mariano
  resultados_df_diebold_mariano_list_ARIMA <- do.call(rbind.data.frame, diebold_mariano_list_ARIMA)
  resultados_df_diebold_mariano_list_VAR <- do.call(rbind.data.frame, diebold_mariano_list_VAR)
  resultados_df_diebold_mariano_list_VAR_ARIMA <- do.call(rbind.data.frame, diebold_mariano_list_VAR_ARIMA)

  colnames(resultados_df_diebold_mariano_list_ARIMA) <- "p-value"
  colnames(resultados_df_diebold_mariano_list_VAR) <- "p-value"
  colnames(resultados_df_diebold_mariano_list_VAR_ARIMA) <- "p-value"

  if (plot) {
    ## MAE
    ggplot(melt(resultados_df[,1:3]), aes(value, fill = variable)) + geom_density(alpha = 0.2) + ggtitle("Distribution : MAE") + scale_x_continuous(limits = c(0, 2))
    ## MSE
    ggplot(melt(resultados_df_MSE[,1:3]), aes(value, fill = variable)) + geom_density(alpha = 0.2) + ggtitle("Distribution : MSE") + scale_x_continuous(limits = c(0, 2))
  }



  saveRDS(resultados_df, file = paste0("MAE/MAE & ",label_to_save,".rds"))

  saveRDS(resultados_df_MSE, file = paste0("MSE/MSE & ",label_to_save,".rds"))

  saveRDS(resultados_df_diebold_mariano_list_ARIMA, file = paste0("DM/DM ARIMA & ",label_to_save,".rds"))

  saveRDS(resultados_df_diebold_mariano_list_VAR, file = paste0("DM/DM VAR & ",label_to_save,".rds"))

  saveRDS(resultados_df_diebold_mariano_list_VAR_ARIMA, file = paste0("DM/DM VAR ARIMA & ",label_to_save,".rds"))

}

CasoReal/codeReal.R
# Libraries
library(stats)
require(tsDyn)
library(vars)
library(Metrics)
library(data.table)
library(urca)
library(tseries)
library(forecast)
library(plotly)

setwd("C:/Users/thiag/Documents/1 -Mestrado/SprintFinal/CasoReal/")

df__ipca <- readxl::read_xlsx("data.xlsx",sheet = "IPCA")
df__usd <- readxl::read_xlsx("data.xlsx",sheet = "USD")

selected_colums <- c("date","Atual","Price")
df__merge <- merge(x = df__ipca, y = df__usd, by = "date", all.x = TRUE)[,selected_colums]
names(df__merge) <- c("date","IPCA","USDBRL")

pre_filter_start <- "2000-01-01"
pre_filter_end <- "2022-02-01"

df__train = df__merge[(df__merge$date >= pre_filter_start) & (df__merge$date <= pre_filter_end),]

p <- plot_ly(df__train) %>%
  add_trace(x = ~date, y = ~IPCA,mode = "lines",  name = "IPCA") %>%  
  add_trace(x = ~date, y = ~USDBRL, mode = "lines", yaxis = "y2", name = "USDBRL") %>%
  layout(yaxis2 = list(overlaying = "y", side = "right"))
plot(p)


scale_value = max(df__train$IPCA, na.rm = TRUE)/ max(df__train$USDBRL, na.rm = TRUE)
ggplot(df__train) + 
  geom_line(aes(x= date, y= IPCA), color = 'dodgerblue') +
  geom_line(aes(x= date, y = USDBRL*scale_value), stat = 'identity') + 
  scale_y_continuous(sec.axis = sec_axis(~./scale_value)) +


df__train_ts <- xts(df__train[,c("IPCA","USDBRL")],order.by=df__train$date)

df__train_ts <- ts(df__train[,c("IPCA","USDBRL")])

df__train_ts <- ts(df__train[,c("IPCA","USDBRL")],order.by=df__train$date)

plot.ts(df__train_ts)



bootstrap.R
library(lattice)
library(latticeExtra)
library(ggplot2)
library(reshape2)
library(boot)

setwd("C:/Users/thiag/Documents/1 -Mestrado/SprintFinal/Defesa/Results/")

for (err in c("MAE","MSE")) {
  
  boot_list <- list()
  
  
  path <- paste0(err,"/")
  
  filenames <- list.files(path, pattern="*rds")
  
  for (file in filenames) {
    
    title <- sub('\\.rds$', '', file)
    
    readed_file <- readRDS(paste0(path,file), refhook = NULL)
    
    base_split <- strsplit(file, split = "&")[[1]]
    par__nobs <- base:::strsplit(base_split[2], split = "__")[[1]][2]
    par__n_ma <- base:::strsplit(base_split[3], split = "__")[[1]][2]
    par__n_fold_ts <- base:::strsplit(base_split[4], split = "__")[[1]][2]
    par__n_ts <- substr(base:::strsplit(base_split[5], split = "__")[[1]][2],1,1)
    
    
    ci_list <- list()
    
    for (modelo_ci in c("ARIMA","VAR","VEC")) {
      boot_fun <- function(data, indices,col){
        d <- data[indices,c(col)]
        mean(d)
      }
      
      boot_sample <- boot(readed_file, boot_fun,col=modelo_ci, R=1000)
      # get 95% confidence interval
      boot_sample_ci <- boot.ci(boot_sample, type="bca",conf = 0.95)
      
      mean_ci <- boot_sample_ci$t0
      sup_ci <- boot_sample_ci$bca[,4]
      inf_ci <- boot_sample_ci$bca[,5]
      
      
      boot_list[[paste0(file,modelo_ci)]] <- list("file"=file,"n_obs"=par__nobs,"n_ma"=par__n_ma,"n_fold_ts"=par__n_fold_ts,"n_ts"=par__n_ts,"model"=modelo_ci,"mean"=mean_ci[1],"sup"=as.numeric(sup_ci),"inf"=as.numeric(inf_ci))
      
      ci_list[[modelo_ci]] <- list(model=modelo_ci,"mean"=mean_ci[1],"sup"=as.numeric(sup_ci),"inf"=as.numeric(inf_ci))
      
    }
    
    ci_df <- do.call(rbind.data.frame, ci_list)
    
    
    
    
    # Chart
    
    png(file=paste0("consolidado/",err,"/boot_charts/",title,".png"),width=1920, height=1080)
    
    plots__ <- ggplot(ci_df) +
      geom_bar( aes(x=model, y=mean), stat="identity", fill="skyblue", alpha=0.7) +
      geom_errorbar( aes(x=model, ymin=inf, ymax=sup), width=0.4, colour="black", alpha=0.9, size=0.9)+
      ggtitle(paste0("Average ",err, " | With Empirical Confidence Intervals")) +
      ylab(err)
    
    plot(plots__)
    
    dev.off()
    
    
    
  }
  
  boot_consolidado <- do.call(rbind.data.frame, boot_list)
  
  writexl::write_xlsx(boot_consolidado,paste0("consolidado/",err,"/boot.xlsx"))
  
}


